### PR TITLE
Pass parsed arguments directly to RCmd

### PR DESCRIPTION
**Pull request by ret2libc**
**Your checklist for this pull request**
- [x] I've read the [guidelines for contributing](https://github.com/radareorg/radare2/blob/master/DEVELOPERS.md) to this repository
- [x] I made sure to follow the project's [coding style](https://github.com/radareorg/radare2/blob/master/DEVELOPERS.md#code-style)
- [x] I've added tests that prove my fix is effective or that my feature works (if possible)
- [ ] I've updated the documentation and the [radare2 book](https://github.com/radareorg/radare2book) with the relevant information (if needed)

**Detailed description**

Right now the new shell just parse the arguments and then put them back together to call `r_cmd_call`. However, in the future  our command handlers (defined in RCmd) will receive argc/argv, so we need RCmd to be able to merge argv together only when needed (that is when we want to call an "old" command handler that does its own parsing).

This PR introduces RCmdParsedArgs and moves the responsability of creating a single "exec string" to RCmd and not in arged_command function.

### CODE REVIEW

**Comment by ret2libc** [comment]
@karliss I tried to split commits in a logic way, so you could try to review them commit-by-commit. Hope it helps.

File: libr/core/cmd.c [hunk 1]
```diff
@@ -4877,51 +4857,27 @@ static char *do_handle_ts_unescape_arg(struct tsr2cmd_state *state, TSNode arg)
...
+			free (unescaped_args[i]);
 		}
-		a->argv_str = create_argv_str (a);
-		return a;
+		free (unescaped_args);
+		return res;
 	} else {
-		return handle_ts_unescape_arg (state, args);
+		char *unescaped_args[] = { do_handle_ts_unescape_arg (state, args) };
+		return r_cmd_parsed_args_newargs (1, unescaped_args);
```
Comment (ret2libc) [inline libr/core/cmd.c:4879]: this probably leaks, i should free unescaped_args[0].

File: libr/core/cmd.c [hunk 2]
```diff
@@ -5172,7 +5165,27 @@ DEFINE_HANDLE_TS_FCN(help_command) {
...
+		RCmdParsedArgs *pr_args = NULL;
+		int res = -1;
+		if (!ts_node_is_null (args)) {
+			pr_args = ts_node_handle_arg_prargs (state, node, args, 1);
+			if (!pr_args) {
+				goto err_else;
+			}
+			r_cmd_parsed_args_setcmd (pr_args, command_str);
+		} else {
+			pr_args = r_cmd_parsed_args_newcmd (command_str);
```
Comment (karliss) [inline libr/core/cmd.c:5180]: Doesn't this need has_space_after_cmd handling?

Comment (ret2libc) [reply libr/core/cmd.c:5180]: no, i'm handling this in exec_string so it does not print the space if there are no additional args apart from argv[0].

File: libr/core/cmd.c [hunk 3]
```diff
@@ -4986,39 +4992,26 @@ DEFINE_HANDLE_TS_FCN(arged_command) {
...
+		}
+	}
+
+	int i;
+	const char *s;
+	r_cmd_parsed_args_foreach_arg (pr_args, i, s) {
+		R_LOG_DEBUG ("parsed_arg %d: '%s'\n", i, s);
 	}
 
 	pr_args->has_space_after_cmd = !ts_node_is_null (args) && ts_node_end_byte (command) < ts_node_start_byte (args);
```
[the remark this answers is not in the document]
Comment (ret2libc) [reply libr/core/cmd.c:5013]: yes, I think I'm handling those with the second part of the `&&` `ts_node_end_byte (command) < ts_node_start_byte (args);`.

So, `arged_command` is composed of `<command> + <args>`. So if  `<command>` is ending before the start of `<args>` it means there is at least a space. If instead it is not, it means there is no space.

[the remark this answers is not in the document]
Comment (karliss) [reply libr/core/cmd.c:5013]: Could you show me an example of command with args that don't have space

[the remark this answers is not in the document]
Comment (ret2libc) [reply libr/core/cmd.c:5013]: https://github.com/radareorg/radare2/pull/16703/files#diff-6165d904c738c087415c0aeca0e91d17R40

`.` is the command. The rest `dr*` is the argument (which is parsed as a command indeed, but then we just need to go back to string).

File: libr/core/cmd.c (hunk 2 above)
Comment (karliss) [reply libr/core/cmd.c:5180]: Then what about the case with args and no space?

Comment (ret2libc) [reply libr/core/cmd.c:5182]: That's handled at https://github.com/radareorg/radare2/pull/16703#discussion_r413905943

Comment (karliss) [reply libr/core/cmd.c:5182]: How does handling it for completely different command "arged_command" help handling it for "help_command"?

**Pull request merged**